doc/source/data/transforming-data.rst

```diff
@@ -1,286 +1,291 @@
+Ray Data represents batches as dicts of NumPy ndarrays or pandas DataFrames. By
+default, Ray Data represents batches as dicts of NumPy ndarrays.
 
-.. literalinclude:: ./doc_code/transforming_data.py
-  :language: python
-  :start-after: __writing_generator_udfs_begin__
-  :end-before: __writing_generator_udfs_end__
+To configure the batch type, specify ``batch_format`` in
+:meth:`~ray.data.Dataset.map_batches`. You can return either format from your function.
 
-------------------
-Shuffle transforms
-------------------
+.. tab-set::
 
-Shuffle transforms change the organization of the data, e.g., increasing the number of blocks, or the order of records in each block, without changing the record contents.
+    .. tab-item:: NumPy
 
-Repartitioning data
-===================
+        .. testcode::
 
-Call :meth:`Dataset.repartition() <ray.data.Dataset.repartition>` to change the
-number of blocks of the dataset. This may be useful to break up your dataset into small
-pieces to enable more fine-grained parallelization, or to reduce the number of files
-produced as output of a write operation.
+            from typing import Dict
+            import numpy as np
+            import ray
 
-.. literalinclude:: ./doc_code/transforming_data.py
-  :language: python
-  :start-after: __shuffle_begin__
-  :end-before: __shuffle_end__
+            def increase_brightness(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+                batch["image"] = np.clip(batch["image"] + 4, 0, 255)
+                return batch
 
-Random shuffle
-==============
+            ds = (
+                ray.data.read_images("example://image-datasets/simple")
+                .map_batches(increase_brightness, batch_format="numpy")
+            )
 
-Call :meth:`Dataset.random_shuffle() <ray.data.Dataset.random_shuffle>` to
-globally shuffle the order of data records.
+    .. tab-item:: pandas
 
-.. doctest::
+        .. testcode::
 
-    >>> import ray
-    >>> dataset = ray.data.range(10)
-    >>> dataset.random_shuffle().take_batch()  # doctest: +SKIP
-    {'id': array([7, 0, 9, 3, 5, 1, 4, 2, 8, 6])}
+            import pandas as pd
+            import ray
 
-For reduced overhead during training ingest, use local shuffles. Read 
-:ref:`Shuffling Data <air-shuffle>` in the AIR user guide to learn more.
+            def drop_nas(batch: pd.DataFrame) -> pd.DataFrame:
+                return batch.dropna()
 
-.. _data-groupbys:
+            ds = (
+                ray.data.read_csv("s3://anonymous@air-example-data/iris.csv")
+                .map_batches(drop_nas, batch_format="pandas")
+            )
 
-------------------
-Grouped transforms
-------------------
+Transforming groups
+===================
```

Comment:
rename to Groupby and transforming groups?

groupby is very common parlance. But transforming groups is more ambiguous.